File: src/main/java/red/tetracube/iot/smartigloohub/messagingbroker/MessagingBrokerService.java
package red.tetracube.iot.smartigloohub.messagingbroker;

import io.smallrye.mutiny.Uni;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import red.tetracube.iot.smartigloohub.configuration.properties.SmartIglooProperties;
import red.tetracube.iot.smartigloohub.data.entities.Switcher;
import red.tetracube.iot.smartigloohub.data.repositories.SwitcherRepository;
import red.tetracube.iot.smartigloohub.data.repositories.TrustedBrokerClientRepository;

import javax.enterprise.context.ApplicationScoped;
import javax.inject.Inject;
import java.util.List;
import java.util.Objects;

@ApplicationScoped
public class MessagingBrokerService {

    @Inject
    SwitcherRepository switcherRepository;

    @Inject
    SmartIglooProperties smartIglooProperties;

    @Inject
    TrustedBrokerClientRepository trustedBrokerClientRepository;

    private final BCryptPasswordEncoder passwordEncoder = new BCryptPasswordEncoder();

    private final static Logger LOGGER = LoggerFactory.getLogger(MessagingBrokerService.class);

    Uni<Void> updateOnlineStatus(String clientId, String action) {
        var wentOnline = List.of("session_terminated", "client_disconnected", "client_connack", "client_connected").contains(action);

        var switcherUni = this.switcherRepository.getSwitcherByClientId(clientId);
        var deviceCombineUni = Uni.combine().all().unis(switcherUni, Uni.createFrom().item(false))
                .asTuple();
        return deviceCombineUni
                .map(devicesTuple -> {
                            var switcher = devicesTuple.getItem1();
                            var other = devicesTuple.getItem2();
                            if (switcher != null && other == null) {
                                LOGGER.info("Switcher found with id ${switcher.id}");
                                return switcher;
                            } else if (switcher == null && other != null) {
                                return other;
                            } else {
                                return null;
                            }
                        }
                )
                .flatMap(device -> {
                    if (device instanceof Switcher switcher) {
                        LOGGER.info("Updating switch with status $wentOnline");
                        switcher.setOnline(wentOnline);
                        return this.switcherRepository.updateSwitcher(switcher);
                    } else {
                        return Uni.createFrom().voidItem();
                    }
                });
    }

    public Uni<Boolean> tryToAuthenticateFeature(String clientId,
                                                 String username,
                                                 String password) {
        LOGGER.info("Checking if authentication is application service");
        if (Objects.equals(password, smartIglooProperties.iot().applicationPassword())) {
            return Uni.createFrom().item(true);
        }

        LOGGER.info("Trying to authenticate feature with client id {} and username {}", clientId, username);
        var trustedBrokerClientUni = trustedBrokerClientRepository.getBrokerClientByClientIdAndUsername(
                clientId,
                username
        );
        return trustedBrokerClientUni.map(brokerClient -> {
            if (brokerClient == null) {
                return false;
            } else {
                return passwordEncoder.matches(password, brokerClient.getPassword());
            }
        });
    }
}

File: src/main/kotlin/red/tetracube/iot/smartigloohub/messagingbroker/api/MessagingBrokerService.kt
package red.tetracube.iot.smartigloohub.messagingbroker.api

import io.smallrye.mutiny.Uni
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.util.*
import javax.enterprise.context.ApplicationScoped

@ApplicationScoped
class MessagingBrokerService {

    companion object {
        private val LOGGER: Logger = LoggerFactory.getLogger(MessagingBrokerService::class.java)
    }

    fun updateOnlineStatus(clientId: UUID, username: String, action: String): Uni<Void> {
        val wentOnline = when (action) {
            "session_terminated", "client_disconnected" -> {
                false
            }
            "client_connack", "client_connected" -> {
                true
            }
            else -> {
                null
            }
        } ?: return Uni.createFrom().voidItem()
     /*   val ledFeature = ledFeatureRepository.getFeatureForAuthenticate(clientId, username)
        val switchFeature = switchFeatureRepository.getFeatureForAuthenticate(clientId, username)
        if (ledFeature == null && switchFeature == null) {
            LOGGER.warn("Cannot find any feature with given client id and username")
        } else if (ledFeature != null && switchFeature == null) {
            LOGGER.info("Checking password for light effect")
            val updatedLedFeature = ledFeature.copy(isOnline = wentOnline)
            return ledFeatureRepository.update(updatedLedFeature)
        } else if (ledFeature == null && switchFeature != null) {
            LOGGER.info("Checking password for the switch")
            val updatedSwitchFeature = switchFeature.copy(isOnline = wentOnline)
            return switchFeatureRepository.update(updatedSwitchFeature)
        }*/
    }
}